File: backend/src/models/saved-search.ts
import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  UpdateDateColumn,
  CreateDateColumn,
  BaseEntity,
  ManyToOne
} from 'typeorm';
import { Vulnerability } from './vulnerability';
import { User } from './user';

@Entity()
export class SavedSearch extends BaseEntity {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column()
  name: string;

  @Column()
  searchTerm: string;

  @Column()
  sortDirection: string;

  @Column()
  sortField: string;

  @Column()
  count: number;

  @Column({ type: 'jsonb', default: '[]' })
  filters: { field: string; values: any[]; type: string }[];

  @Column()
  searchPath: string;

  @ManyToOne((type) => User, {
    onDelete: 'SET NULL',
    onUpdate: 'CASCADE'
  })
  createdBy: User;
}

File: backend/src/xfd_django/xfd_api/schema_models/saved_search.py
"""Saved Search schemas."""
# Standard Python Libraries
from datetime import datetime
from typing import Any, List, Optional
from uuid import UUID

# Third-Party Libraries
from pydantic import BaseModel


class SavedSearchFilters(BaseModel):
    """SavedSearchFilters schema."""

    type: Optional[str]
    field: str
    values: List[Any]


class SavedSearchCreate(BaseModel):
    """Saved search create."""

    name: str
    search_term: str
    sort_direction: str
    sort_field: str
    count: int
    filters: List[SavedSearchFilters]
    search_path: str


class SavedSearchUpdate(BaseModel):
    """Saved search update."""

    name: str
    search_term: str
    sort_direction: str
    sort_field: str
    count: int
    filters: List[SavedSearchFilters]
    search_path: str


class SavedSearch(BaseModel):
    """SavedSearch schema."""

    id: UUID
    created_at: datetime
    updated_at: datetime
    name: str
    search_term: str
    sort_direction: str
    sort_field: str
    count: int
    filters: List[SavedSearchFilters]
    search_path: str
    created_by_id: UUID


class SavedSearchList(BaseModel):
    """SavedSearchList schema."""

    result: List[SavedSearch]
    count: int
